### PR TITLE
optimize null flags encoding (#757)

### PR DESCRIPTION
**Pull request by DeepThinker666**
use lz4 to compress null flags for better compression ratio and decompression speed. In my test, Bitshuffle has no compression ratio for null flags.  In a test with 266882552 records, the compressed data size and decompression time is as following:
                  bitshuffle          lz4          rle
size            268MB             30MB      20MB
time            246ms             119ms      262ms
Add config::null_encoding to decide the default encoding type of null flags. Make sure to upgrade all BEs before change
the config.

### CODE REVIEW

**Review by sduzh — DISMISSED** [review]

File: be/src/storage/rowset/segment_v2/column_writer.cpp [hunk 1]
```diff
@@ -546,7 +568,7 @@ Status ScalarColumnWriter::finish_current_page() {
         size_t num_null = data_page_footer->num_values() - num_data;
         // If more than 80% of the current page is NULL records, using format 1 for the next page,
         // otherwise using format 2.
-        _curr_page_format = (num_null > 4 * num_data) ? 1 : 2;
+        _curr_page_format = (num_null > 4 * num_data) ? 1 : config::default_page_format;
```
Comment (chaoyli) [inline be/src/storage/rowset/segment_v2/column_writer.cpp:571]: If one be is page_format==3, another be is page_format == 2, how to handle the clone logic?

File: gensrc/proto/segment_v2.proto [hunk 2]
```diff
@@ -77,6 +77,9 @@ message DataPageFooterPB {
     // another difference is that the format 1 use Run-Length encoding to encode the null map,
     // while format 2 use the bitshuffle.
     optional uint32 format_version = 20;
+    // 0 for bitshuffle
+    // 1 for lz4
+    optional uint32 null_flag_version = 21;
```
Comment (imay) [inline gensrc/proto/segment_v2.proto:82]: 1. better to rename it to null_format
2. enum is preferred

File: be/src/storage/rowset/segment_v2/parsed_page.cpp [hunk 3]
```diff
@@ -295,16 +296,41 @@ Status parse_page_v2(std::unique_ptr<ParsedPage>* result, PageHandle handle, con
...
-        int64_t r =
-                bitshuffle::decompress_lz4(null_flags.data, page->_null_flags.data(), elements_pad, sizeof(uint8_t), 0);
-        if (r < 0) {
-            return Status::Corruption("bitshuffle decompress failed: " + bitshuffle_error_msg(r));
+        LOG(INFO) << "footer.has_null_flag_version():" << footer.has_null_flag_version();
+        std::cout << "footer.has_null_flag_version():" << footer.has_null_flag_version() << std::endl;
+        if (footer.has_null_flag_version()) {
+            LOG(INFO) << "footer.null_flag_version():" << footer.null_flag_version();
+            std::cout << "footer.null_flag_version():" << footer.null_flag_version() << std::endl;
+        }
```
Comment (imay) [inline be/src/storage/rowset/segment_v2/parsed_page.cpp:304]: these log should be removed.

File: be/src/storage/rowset/segment_v2/parsed_page.cpp [hunk 4]
```diff
@@ -295,16 +296,41 @@ Status parse_page_v2(std::unique_ptr<ParsedPage>* result, PageHandle handle, con
...
+        uint32_t null_flag_version = footer.has_null_flag_version() ? footer.null_flag_version() : 0;
+        if (null_flag_version == 0) {
+            LOG(INFO) << "parse null flag use bitshuffle";
+            std::cout << "parse null flag use bitshuffle" << std::endl;
```
Comment (imay) [inline be/src/storage/rowset/segment_v2/parsed_page.cpp:308]: No debug log

File: be/src/storage/rowset/segment_v2/column_writer.cpp [hunk 5]
```diff
@@ -523,6 +555,11 @@ Status ScalarColumnWriter::finish_current_page() {
     data_page_footer->set_nullmap_size(nullmap.slice().size);
     data_page_footer->set_format_version(_curr_page_format);
     data_page_footer->set_corresponding_element_ordinal(_element_ordinal);
+    if (is_nullable() && (_curr_page_format == 2)) {
+        LOG(INFO) << "set null flag version in footer:" << _null_map_builder_v2->null_flag_version();
+        std::cout << "set null flag version in footer:" << _null_map_builder_v2->null_flag_version() << std::endl;
```
Comment (imay) [inline be/src/storage/rowset/segment_v2/column_writer.cpp:560]: Log should be removed

**Review by imay — APPROVED** [review]

**Review by sduzh — APPROVED** [review]

**Pull request merged**